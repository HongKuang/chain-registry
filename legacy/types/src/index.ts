export { Asset, AssetList } from './assetlist.schema';
export { Chain } from './chain.schema';
// temporarily use IBCInfo until v2
export { IBCData as IBCInfo } from './ibc_data.schema';
export * from './registry';
