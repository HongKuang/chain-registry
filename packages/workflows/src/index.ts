export * from './fixtures';
export * from './schema-typescript';