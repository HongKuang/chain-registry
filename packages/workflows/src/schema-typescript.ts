import { basename, dirname, join } from 'path';
import { generateTypeScript, JSONSchema, SchemaTSOptions } from 'schema-typescript';

// Define an interface for the filesystem operations
export interface FileSystem {
  readFileSync(path: string, encoding: string): string;
  writeFileSync(path: string, data: string): void;
  mkdirpSync(path: string): void;
}

// Define an interface for the constructor options
export interface SchemaTypeGeneratorOptions {
  outputDir: string;
  schemaTSOptions: Partial<SchemaTSOptions>;
  readFs: FileSystem;
  writeFs: FileSystem;
  schemas: string[];
  supportedSchemas?: string[];
}

// Default titles for certain schemas
// TODO create issue in cosmos/chain-registry
const DEFAULT_TITLES: { [filename: string]: string } = {
  'ibc_data.schema.json': 'IBCData',
  'chain.schema.json': 'Chain',
  'assetlist.schema.json': 'AssetList'
};
export class SchemaTypeGenerator {
  private outputDir: string;
  private writeFs: FileSystem;
  private readFs: FileSystem;
  private schemas: string[];
  private schemaTSOptions: Partial<SchemaTSOptions>
  private supportedSchemas: Set<string>;

  constructor(options: SchemaTypeGeneratorOptions) {
    this.outputDir = options.outputDir;
    this.writeFs = options.writeFs;
    this.readFs = options.readFs;
    this.schemas = options.schemas;
    this.supportedSchemas = new Set(options.supportedSchemas || []);
  }

  private updateSchemaTitle(schema: JSONSchema, schemaFile: string): void {
    const filename = basename(schemaFile);
    if (DEFAULT_TITLES[filename]) {
      schema.title = DEFAULT_TITLES[filename];
    }
  }

  private isSchemaSupported(filePath: string): boolean {
    if (this.supportedSchemas.size === 0) return true; // If no filter is provided, process all files
    const filename = basename(filePath);
    return this.supportedSchemas.has(filename);
  }

  public generateTypes(): void {
    this.schemas.forEach(schemaFile => {
      if (this.isSchemaSupported(schemaFile)) {
        try {
          const schema: JSONSchema = this.readJsonFile(schemaFile);
          this.updateSchemaTitle(schema, schemaFile);
          const result = this.generateTypeScript(schema);
          const filename = this.getOutputFilename(schemaFile);
          this.ensureDirExists(filename);
          this.writeFile(filename, result);
        } catch (e) {
          console.log({ schemaFile })
        }
      }
    });
  }

  private readJsonFile(filePath: string): JSONSchema {
    return JSON.parse(this.readFs.readFileSync(filePath, 'utf-8'));
  }

  private generateTypeScript(schema: JSONSchema): string {
    return generateTypeScript(schema, this.schemaTSOptions);
  }

  private getOutputFilename(schemaFile: string): string {
    const filename = basename(schemaFile);
    return join(this.outputDir, filename.replace(/.json$/, '.ts'));
  }

  private ensureDirExists(filePath: string): void {
    this.writeFs.mkdirpSync(dirname(filePath));
  }

  private writeFile(filePath: string, content: string): void {
    this.writeFs.writeFileSync(filePath, content);
  }
}
