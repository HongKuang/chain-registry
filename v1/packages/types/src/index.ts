export { Asset, AssetList, DenomUnit } from './assetlist.schema';
export { Chain } from './chain.schema';
// temporarily use IBCInfo until v2
export { ChannelInfo, IBCData as IBCInfo } from './ibc_data.schema';
export * from './registry';
