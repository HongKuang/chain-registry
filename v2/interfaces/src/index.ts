export { Asset, AssetList } from './assetlist.schema';
export { Chain } from './chain.schema';
export { IBCData } from './ibc_data.schema';