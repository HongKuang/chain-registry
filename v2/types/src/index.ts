export * from './assetlist.schema';
export * from './chain.schema';
export * from './ibc_data.schema';
export * from './memo_keys.schema';
export * from './shared';
export * from './versions.schema';