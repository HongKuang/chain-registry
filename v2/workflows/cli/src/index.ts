export * from './commands/info';
export * from './utils';