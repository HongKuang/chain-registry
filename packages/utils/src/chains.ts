import { Chain } from '@chain-registry/types';

import { customFind } from './utils';

export interface GasPriceRanges {
  low: number;
  average: number;
  high: number;
}

export const getGasPriceRangesFromChain = (chain: Chain): GasPriceRanges => {
  const feeToken = chain.fees?.fee_tokens?.[0];
  return {
    low: feeToken?.low_gas_price ?? 0.01,
    average: feeToken?.average_gas_price ?? 0.025,
    high: feeToken?.high_gas_price ?? 0.04
  };
};

export const getChainByChainName = (
  chains: Chain[],
  chainName: string
): Chain | undefined => {
  return customFind(chains, (chain) => chain.chain_name === chainName, `chainName:${chainName}`);
};

export const getChainByChainId = (
  chains: Chain[],
  chainId: string
): Chain | undefined => {
  return customFind(chains, (chain) => chain.chain_id === chainId, `chainId:${chainId}`);
};

export const getChainNameByChainId = (
  chains: Chain[],
  chainId: string
): string | undefined => {
  return getChainByChainId(chains, chainId)?.chain_name;
};

export const getChainIdByChainName = (
  chains: Chain[],
  chainName: string
): string | undefined => {
  return getChainByChainName(chains, chainName)?.chain_id;
};

export const getChainGasPriceRanges = (
  chains: Chain[],
  chainName: string
): GasPriceRanges | undefined => {
  const chain = getChainByChainName(chains, chainName);
  return chain ? getGasPriceRangesFromChain(chain) : undefined;
};

export const getChainPrettyName = (
  chains: Chain[],
  chainName: string
): string | undefined => {
  return getChainByChainName(chains, chainName)?.pretty_name;
};

export const getChainBech32Prefix = (
  chains: Chain[],
  chainName: string
): string | undefined => {
  return getChainByChainName(chains, chainName)?.bech32_prefix;
};
