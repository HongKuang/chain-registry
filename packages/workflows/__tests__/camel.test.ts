import { fs } from 'file-ts';
import * as nativeFs from 'fs';

import { schemas } from '../src';
import { SchemaTypeGenerator } from '../src';
import { globSync, mkdirpSync } from '../test-utils';

it('types', () => {
  const generator = new SchemaTypeGenerator({
    outputDir: './schemas',
    // @ts-ignore
    readFs: nativeFs,
    // @ts-ignore
    writeFs: {
        ...fs,
        mkdirpSync
    },
    supportedSchemas: [
        'chain.schema.json',
        'assetlist.schema.json',
        // 'ibc_data.schema.json'
    ],
    schemas,
    schemaTSOptions: {
        strictTypeSafety: true,
        useCamelCase: true,
        useSingleQuotes: true
    }
  });
  generator.generateTypes();

  const tsFiles = globSync('./schemas', '*.ts');
  expect(tsFiles).toMatchSnapshot();

  tsFiles.forEach(file=>{
    const content = fs.readFileSync(file, 'utf-8');
    expect(content).toMatchSnapshot();
  });
  
});