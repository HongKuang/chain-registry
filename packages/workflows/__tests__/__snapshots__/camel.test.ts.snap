// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`types 1`] = `
[
  "schemas/assetlist.schema.ts",
  "schemas/chain.schema.ts",
]
`;

exports[`types 2`] = `
"export interface Asset {
  deprecated?: boolean;
  description?: string;
  extended_description?: string;
  denom_units: DenomUnit[];
  type_asset?: "sdk.coin" | "cw20" | "erc20" | "ics20" | "snip20" | "snip25" | "bitcoin-like" | "evm-base" | "svm-base" | "substrate";
  address?: string;
  base: string;
  name: string;
  display: string;
  symbol: string;
  traces?: (IbcTransition | IbcCw20Transition | NonIbcTransition)[];
  ibc?: {
    source_channel: string;
    dst_channel: string;
    source_denom: string;
  };
  logo_URIs?: {
    png?: string;
    svg?: string;
  };
  images?: {
    image_sync?: Pointer;
    png?: string;
    svg?: string;
    theme?: {
      primary_color_hex?: string;
      circle?: boolean;
      dark_mode?: boolean;
    };
  }[];
  coingecko_id?: string;
  keywords?: string[];
  socials?: {
    website?: string;
    twitter?: string;
  };
}
export interface DenomUnit {
  denom: string;
  exponent: number;
  aliases?: string[];
}
export interface Pointer {
  chain_name: string;
  base_denom?: string;
}
export interface IbcTransition {
  type: "ibc";
  counterparty: {
    chain_name: string;
    base_denom: string;
    channel_id: string;
  };
  chain: {
    channel_id: string;
    path: string;
  };
}
export interface IbcCw20Transition {
  type: "ibc-cw20";
  counterparty: {
    chain_name: string;
    base_denom: string;
    port: string;
    channel_id: string;
  };
  chain: {
    port: string;
    channel_id: string;
    path: string;
  };
}
export interface NonIbcTransition {
  type: "bridge" | "liquid-stake" | "synthetic" | "wrapped" | "additional-mintage" | "test-mintage";
  counterparty: {
    chain_name: string;
    base_denom: string;
    contract?: string;
  };
  chain?: {
    contract: string;
  };
  provider: string;
}
export interface AssetLists {
  $schema?: string;
  chain_name: string;
  assets: Asset[];
}"
`;

exports[`types 3`] = `
"export interface Peer {
  id: string;
  address: string;
  provider?: string;
}
export interface Endpoint {
  address: string;
  provider?: string;
  archive?: boolean;
}
export interface Explorer {
  kind?: string;
  url?: string;
  tx_page?: string;
  account_page?: string;
}
export interface FeeToken {
  denom: string;
  fixed_min_gas_price?: number;
  low_gas_price?: number;
  average_gas_price?: number;
  high_gas_price?: number;
  gas_costs?: {
    cosmos_send?: number;
    ibc_transfer?: number;
  };
}
export interface StakingToken {
  denom: string;
}
export interface Pointer {
  chain_name: string;
  base_denom?: string;
}
export interface CosmosChain {
  $schema?: string;
  chain_name: string;
  chain_id: string;
  pre_fork_chain_name?: string;
  pretty_name?: string;
  website?: string;
  update_link?: string;
  status?: "live" | "upcoming" | "killed";
  network_type?: "mainnet" | "testnet" | "devnet";
  bech32_prefix: string;
  bech32_config?: {
    bech32PrefixAccAddr?: string;
    bech32PrefixAccPub?: string;
    bech32PrefixValAddr?: string;
    bech32PrefixValPub?: string;
    bech32PrefixConsAddr?: string;
    bech32PrefixConsPub?: string;
  };
  daemon_name?: string;
  node_home?: string;
  key_algos?: ("secp256k1" | "ethsecp256k1" | "ed25519" | "sr25519" | "bn254")[];
  slip44?: number;
  alternative_slip44s?: number[];
  fees?: {
    fee_tokens: FeeToken[];
  };
  staking?: {
    staking_tokens: StakingToken[];
    lock_duration?: {
      blocks?: number;
      time?: string;
    };
  };
  codebase?: {
    git_repo?: string;
    recommended_version?: string;
    go_version?: string;
    compatible_versions?: string[];
    binaries?: {
      "linux/amd64"?: string;
      "linux/arm64"?: string;
      "darwin/amd64"?: string;
      "darwin/arm64"?: string;
      "windows/amd64"?: string;
      "windows/arm64"?: string;
    };
    cosmos_sdk_version?: string;
    consensus?: {
      type: "tendermint" | "cometbft" | "sei-tendermint";
      version?: string;
    };
    cosmwasm_version?: string;
    cosmwasm_enabled?: boolean;
    cosmwasm_path?: string;
    ibc_go_version?: string;
    ics_enabled?: ("ics20-1" | "ics27-1" | "mauth")[];
    genesis?: {
      name?: string;
      genesis_url: string;
      ics_ccv_url?: string;
    };
    versions?: {
      name: string;
      tag?: string;
      height?: number;
      proposal?: number;
      previous_version_name?: string;
      next_version_name?: string;
      recommended_version?: string;
      go_version?: string;
      compatible_versions?: string[];
      cosmos_sdk_version?: string;
      consensus?: {
        type: "tendermint" | "cometbft" | "sei-tendermint";
        version?: string;
      };
      cosmwasm_version?: string;
      cosmwasm_enabled?: boolean;
      cosmwasm_path?: string;
      ibc_go_version?: string;
      ics_enabled?: ("ics20-1" | "ics27-1" | "mauth")[];
      binaries?: {
        "linux/amd64"?: string;
        "linux/arm64"?: string;
        "darwin/amd64"?: string;
        "darwin/arm64"?: string;
        "windows/amd64"?: string;
        "windows/arm64"?: string;
      };
    }[];
  };
  images?: {
    image_sync?: Pointer;
    png?: string;
    svg?: string;
    theme?: {
      primary_color_hex?: string;
      circle?: boolean;
      dark_mode?: boolean;
    };
    layout?: "logo" | "logomark" | "logotype";
    text_position?: "top" | "bottom" | "left" | "right" | "integrated";
  }[];
  logo_URIs?: {
    png?: string;
    svg?: string;
  };
  description?: string;
  peers?: {
    seeds?: Peer[];
    persistent_peers?: Peer[];
  };
  apis?: {
    rpc?: Endpoint[];
    rest?: Endpoint[];
    grpc?: Endpoint[];
    wss?: Endpoint[];
    "grpc-web"?: Endpoint[];
    "evm-http-jsonrpc"?: Endpoint[];
  };
  explorers?: Explorer[];
  keywords?: string[];
  extra_codecs?: ("ethermint" | "injective")[];
}"
`;
